tests/queries/0_stateless/02456_test_attach_partition_different_partition_exp.sql

```diff
@@ -0,0 +1,154 @@
+-- { echoOn }
+-- Should be allowed since destination partition expr is monotonically increasing and compatible
+DROP TABLE IF EXISTS source;
+DROP TABLE IF EXISTS destination;
+
+CREATE TABLE source (timestamp DateTime) engine=MergeTree ORDER BY tuple() PARTITION BY toYYYYMMDD(timestamp);
+CREATE TABLE destination (timestamp DateTime) engine=MergeTree ORDER BY tuple() PARTITION BY toYYYYMM(timestamp);
+
+INSERT INTO TABLE source VALUES ('2010-03-02 02:01:01'), ('2010-03-02 02:01:03');
+
+ALTER TABLE destination ATTACH PARTITION ID '20100302' FROM source;
+
+SELECT * FROM source ORDER BY timestamp;
+SELECT * FROM destination ORDER BY timestamp;
+SELECT partition_id FROM system.parts where table='destination' AND database = currentDatabase();
+
+-- Should be allowed since destination partition expr is monotonically increasing and compatible. Note that even though
+-- the destination partition expression is more granular, the data would still fall in the same partition. Thus, it is valid
+DROP TABLE IF EXISTS source;
+DROP TABLE IF EXISTS destination;
+
+CREATE TABLE source (timestamp DateTime) engine=MergeTree ORDER BY tuple() PARTITION BY toYYYYMM(timestamp);
+CREATE TABLE destination (timestamp DateTime) engine=MergeTree ORDER BY tuple() PARTITION BY toYYYYMMDD(timestamp);
+
+INSERT INTO TABLE source VALUES ('2010-03-02 02:01:01'), ('2010-03-02 02:01:03');
+
+ALTER TABLE destination ATTACH PARTITION ID '201003' FROM source;
+
+SELECT * FROM source ORDER BY timestamp;
+SELECT * FROM destination ORDER BY timestamp;
```

Comment:
Flaky test. You need to add `OPTIMIZE TABLE destination FINAL` or change the engine to MergeTree. SummingMergeTree can merge these two rows into a single one.

Reply:
Everything else is fine.